File: permissionx/src/main/java/com/lws/permissionx/RationaleController.java
package com.lws.permissionx;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.provider.Settings;

import androidx.appcompat.app.AlertDialog;

class RationaleController{
    private final PermissionBuilder permissionBuilder;


    CharSequence requestRationale;

    CharSequence deniedRationale;

    CharSequence deniedForeverRationale;

    Runnable deniedRationaleNegativeListener;

    Runnable deniedForeverRationaleNegativeListener;


    RationaleController(PermissionBuilder permissionBuilder) {
        this.permissionBuilder = permissionBuilder;
    }

    boolean canShowRequestRationale() {
        return requestRationale != null;
    }

    boolean canShowDeniedRationale() {
        return deniedRationale != null && deniedRationaleNegativeListener != null;
    }

    boolean canShowDeniedForeverRationale() {
        return deniedForeverRationale != null && deniedForeverRationaleNegativeListener != null;
    }


    void showRequestRationale() {
        AlertDialog rationaleDialog = new AlertDialog.Builder(permissionBuilder.activity, PermissionX.getDefaultConfig().getAlertDialogTheme())
                .setCancelable(false)
                .setMessage(requestRationale)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> permissionBuilder.getInvisibleFragment().request(permissionBuilder))
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> permissionBuilder.onCancelRationale())
                .create();
        show(rationaleDialog);
    }

    void showDeniedRationale() {
        if (!canShowDeniedRationale()) {
            return;
        }
        AlertDialog rationaleDialog = new AlertDialog.Builder(permissionBuilder.activity, PermissionX.getDefaultConfig().getAlertDialogTheme())
                .setCancelable(false)
                .setMessage(deniedRationale)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> permissionBuilder.getInvisibleFragment().request(permissionBuilder))
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> deniedRationaleNegativeListener.run())
                .create();
        show(rationaleDialog);
    }

    void showDeniedForeverRationale() {
        if (!canShowDeniedForeverRationale()) {
            return;
        }
        AlertDialog rationaleDialog = new AlertDialog.Builder(permissionBuilder.activity, PermissionX.getDefaultConfig().getAlertDialogTheme())
                .setCancelable(false)
                .setMessage(deniedForeverRationale)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> forwardToSettings(permissionBuilder.activity))
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> deniedForeverRationaleNegativeListener.run())
                .create();
        show(rationaleDialog);
    }


    private void show(AlertDialog alertDialog) {
        alertDialog.show();
        alertDialog.getWindow().setGravity(PermissionX.getDefaultConfig().getGravity());
    }

    private void forwardToSettings(Context context) {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        Uri uri = Uri.fromParts("package", context.getPackageName(), null);
        intent.setData(uri);
        context.startActivity(intent);
    }

}

File: permissionx/src/main/java/com/lws/permissionx/RationaleController.kt
package com.lws.permissionx

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import androidx.appcompat.app.AlertDialog
import com.lws.permissionx.PermissionX.defaultConfig

internal class RationaleController(
    private val permissionBuilder: PermissionBuilder,
    private val context: Context,
    private val permissions: Array<out String>
) {
    var requestRationale: CharSequence? = null
    var deniedRationale: CharSequence? = null
    var deniedForeverRationale: CharSequence? = null
    var deniedRationaleNegativeListener: Runnable? = null
    var deniedForeverRationaleNegativeListener: Runnable? = null

    fun hasRequestRationale(): Boolean {
        return requestRationale != null
    }

    private fun hasDeniedRationale(): Boolean {
        return deniedRationale != null
    }

    private fun hasDeniedForeverRationale(): Boolean {
        return deniedForeverRationale != null
    }

    fun showRequestRationale() {
        AlertDialog.Builder(context, defaultConfig.alertDialogTheme)
            .setCancelable(false)
            .setMessage(requestRationale)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, which: Int ->
                permissionBuilder.requestPermissions()
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, which: Int ->
                val result: MutableMap<String, Boolean> = LinkedHashMap()
                for (permission in permissions) {
                    result[permission] = PermissionX.hasPermissions(context, permission)
                }
                permissionBuilder.handleResult(result, true)
            }
            .create()
            .showWithGravity()
    }

    fun showDeniedRationale() {
        if (!hasDeniedRationale()) {
            return
        }
        AlertDialog.Builder(context, defaultConfig.alertDialogTheme)
            .setCancelable(false)
            .setMessage(deniedRationale)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface?, which: Int ->
                permissionBuilder.requestPermissions()
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, which: Int ->
                deniedRationaleNegativeListener?.run()
            }
            .create()
            .showWithGravity()

    }

    fun showDeniedForeverRationale() {
        if (!hasDeniedForeverRationale()) {
            return
        }
        AlertDialog.Builder(context, defaultConfig.alertDialogTheme)
            .setCancelable(false)
            .setMessage(deniedForeverRationale)
            .setPositiveButton(android.R.string.ok) { dialog: DialogInterface, which: Int ->
                Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                    data = Uri.fromParts("package", context.packageName, null)
                }.also {
                    context.startActivity(it)
                }
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface, which: Int ->
                deniedForeverRationaleNegativeListener?.run()
            }
            .create()
            .showWithGravity()

    }


    private fun AlertDialog.showWithGravity() {
        show()
        window?.setGravity(defaultConfig.gravity)
    }
}